### ISSUE TITLE
scroll-top doesn't work as expected inside ng-if

### ISSUE THREAD

**Issue by xak2000** [brantwills/Angular-Paging#38]
If i set `scroll-top="false"` in `pager` directive, then it doesn't work as expected if `pager` directive is inside `ng-if` block.

I slightly modified `pager` template to show current value of `scrollTop` variable.
Open [this plunker](http://plnkr.co/edit/Q3kRsUT12VEWbdrhG7gR?p=preview) and see the value of `scrollTop` variable. It is `"false"` (string value) and because of casting it is `true` as boolean. Because of this after clicking on `[2]` the page is scrolling to the top of course. After this step the `scrollTop` variable is boolean again (not string anymore) and all works as expected. But if you click `Show/hide div` button then it is string again.

I think this is because of isolated scope bindings somehow override the values which you set in `setScopeValues` function. They binded with `@` so are always strings. But in `setScopeValues` you set them as `scope.scrollTop = scope.$eval(attrs.scrollTop);`.

I think the simplest and corrent way to fix this issue is to remove scope bindings and rely on `setScopeValues` function. But for this you need to modify the variable assignment from `scope.scrollTop = scope.$eval(attrs.scrollTop);` to `scope.scrollTop = angular.isDefined(attrs.scrollTop) ? scope.$eval(attrs.scrollTop) : false;` where last `false` can be any default value if attribute is missing in html template.

I think this issue affects not only `scrollTop` but any attribute which assigned through bindings and manually at the same time.

**Comment by brantwills** [comment]
Hey xak2000, great catch!  I've always felt strange about those `scope.$eval()` items, there just has to be a better way more elegant way.. I will look into this as soon as possible (estimating the first week in January) and try to build up some test cases around ng-if and other common directives

Huge thanks again for capturing the issue!
I've added the bug label to this issue

**Comment by xak2000** [comment]
I think you should use
`scope.$parent.$eval(attrs.model);` (`$parent` because of isolated scope)
**or**
isolated scope bindings (`=` in case of scrollTop as it is not a string)
but not both at the same time for the same attribute.

I tested isolated scope bindings and realized that they happens not only *before* controller and link function run but also *after* them. So if user set some attribute in template like `<my-dir my-attr="value">` and this attribute is bound to isolated scope (through `@`) and then in controller or link function of this directive we set another value to scope variable bounded to this attribte, then after link function finish, this variable silently set to attribute value again, losing value setted in controller or link function.

I don't know if it intended behaviour or not, but this is how it works. We just can't set new value to variable that is bound to attribute through `@`. Or rather we can, but can't be sure that bounding mechanism will not override our value in future.

---
What you choose is depends on necessity of watching future attribute changes. Isolated bindings always watching for changes and updates our variables. Manual bindings is more flexible and you must  create watches manually if need.

**Comment by xak2000** [comment]
> there just has to be a more elegant way..

About this. More elegant way I think is just using `=?` bindings.
Something like that:
```
scope: {
  scrollTop: '=?'
}
```
then in directive's controller
```
$scope.scrollTop = $scope.scrollTop || false;
```
But this creates another +1 watch if `scroll-top` attribute is present. I just hate watches. :) I trying to avoid them all ways possible. So I recommend to just use `$eval` for attributes you don't care if they will change in the future. But it is up to you of course!

For myself I defined a rule: If my directive can potentially be used inside `ng-repeat` then avoid watchers as hard as I can.
---
Another possibility is:
```
scope: {
  scrollTop: '&'
}
```
And using it in places where it needed like:
```
$scope.scrollTop() || false;
```
This will not create any watchers but will access parent's scope value through function which will guarantee we can't accidentally change parent value (which we can while using `=` binding).

But I'm not sure if this use of `&` is considered as best practice as not saw it anywhere. All just using `=` and don't care about watchers. :-/

**Comment by brantwills** [comment]
Hey xak2000,

Once again sorry for the delay, last month was crazy, but I finally have some time to spend on this.

Thank you once again for the plunker, it really really helped to get an idea of the issue

I found a nice solution which keeps `@` avoiding watchers and doesn't introduce breaking changes

In short, I noticed we are up against a string vs bool comparison in the if statement around line 204:

```javascript
        // If allowed scroll up to the top of the page
        if (scope.scrollTop) {
            scrollTo(0, 0);
        }
```

To which I've updated to 

```javascript
        // If allowed scroll up to the top of the page
        // We use == 'true' for string and boolean comparison
        if (scope.scrollTop == 'true') {
            scrollTo(0, 0);
        }
```

This corrects the ng-if nested issues without adding an extra watcher, which I agree should be avoided as much as possible!  

I've pushed branch iss38 if you would like to review the code, let me know what you think.  Although still a bit hacky with the loose `==` comparison it seems to get what we need.

I would like to write up some test scripts, once those are complete I will merge these changes into master and close out the issue.

**Comment by xak2000** [comment]
Why not just remove `scrollTop: '@'` binding and modify `setScopeValues` function to set `scope.scrollTop` value as this:
```
scope.scrollTop = angular.isDefined(attrs.scrollTop) ? !!scope.$parent.$eval(attrs.scrollTop) : false;
```

This way `scope.scrollTop` variable is always boolean and we don't need hacky way to compare it with string. Also it would work with all values which can be type-casted to boolean. For example `1` and `0`. Not only `true`. And also it will work even with variable in template. User can use it like
```
      <paging
        page="currentPage"
        page-size="pageSize"
        total="totalElements"
        scroll-top="myScrollTop">
      </paging>
```
and in controller: `$scope.myScrollTop = false;`

I created a new [plunk](http://plnkr.co/edit/073QbZ6mAuljeHdxLIGs?p=preview) to show this.
Maybe I did not consider something.. But it works I think.

**Comment by brantwills** [comment]
Done and Done!

I addition, thought it would be good to also apply this logic to the rest of the boolean settings. This allows us to take the `@` out for the other settings which I think makes things a bit cleaner.  As far as I can tell this does not introduce any breaking changes and all my tests pass! I pushed the code to iss38 branch if interested. 

Again a huuge thanks!!

**Comment by xak2000** [comment]
Yes you are right about other boolean attributes. I also recommend to add third parameter to `evalBoolAttribute(scope, value)` function. Like `evalBoolAttribute(scope, value, defaultValue)` and use it like this:
`scope.hideIfEmpty = evalBoolAttribute(scope, attrs.hideIfEmpty, false);`
This way you can assign any default value (not only `false`) if this attribute is absent in user template.
For exampe if you want `hideIfEmpty` to be `true` if user doesn't pass it from template.

But of course you can do it later if `angular-paging` will really need not falsy default value for some parameter.

Again, thanks you for this good module! Keep up the good work! :+1:

**Comment by brantwills** [comment]
Closed with version 2.2.0 release
